cpp/comms/src/cuML_std_comms_impl.hpp

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include <nccl.h>
+
+#include <common/cuml_comms_iface.hpp>
+
+#ifdef WITH_UCX
+#include <ucp/api/ucp.h>
+struct ucx_context {
+  int completed;
+  bool needs_release = true;
+};
+#endif
+
+namespace ML {
+
+/**
+ * @brief A cumlCommunicator implementation capable of running collective communications
+ * with NCCL and point-to-point-communications with UCX. Note that the latter is optional.
+ *
+ * Underlying comms, like NCCL and UCX, should be initialized and ready for use,
+ * and maintained, outside of the cuML Comms lifecycle. This allows us to decouple the
+ * ownership of the actual comms from cuml so that they can also be used outside of cuml.
+ *
+ * For instance, nccl-py can be used to bootstrap a ncclComm_t before it is
+ * used to construct a cuml comms instance. UCX endpoints can be bootstrapped
+ * in Python using ucx-py, before being used to construct a cuML comms instance.
+ */
+class cumlStdCommunicator_impl : public MLCommon::cumlCommunicator_iface {
+ public:
+  cumlStdCommunicator_impl() = delete;
+
+#ifdef WITH_UCX
+
+  /**
+   * @brief Constructor for collective + point-to-point operation.
+   * @param comm initialized nccl comm
+   * @param ucp_worker initialized ucp_worker instance
+   * @param eps shared pointer to array of ucp endpoints
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  cumlStdCommunicator_impl(ncclComm_t comm, ucp_worker_h ucp_worker,
+                           std::shared_ptr<ucp_ep_h*> eps, int size, int rank);
```

Comment:
I think the user of `cumlStdCommunicator_impl` owning `eps` will always outlive `cumlStdCommunicator_impl` I would therefore consider passing in `eps` as plain pointer. While this would unfortunately the ownership contract regarding `eps` less clear it would allow the simplification of `inject_comms` mentioned above.

Reply:
My knowledge of Cython is limited but I believe I might be able to accomplish what I need w/ ctypes. Going to give it a try.